Sources/Hyperspace/Request/Request.URLRequestCreationStrategy.swift

```diff
@@ -0,0 +1,39 @@
+//
+//  Request.URLRequestCreationStrategy.swift
+//  Hyperspace-iOS
+//
+//  Created by William McGinty on 3/6/20.
+//  Copyright © 2020 Bottle Rocket Studios. All rights reserved.
+//
+
+import Foundation
+
+public extension Request {
+    
+    // MARK: - URLRequestCreationStrategy
+    
+    enum URLRequestCreationStrategy {
```

Comment:
I originally built it as an `enum` for readability at the call site. I figured that at the end of the day, every possible `URLRequestCreationStrategy` can be constructed using the `.custom((Request) -> URLRequest)` case. This would allow me to have 1 case defined on the `enum`, and then `static var/func`(s) in order to define different strategies (for example, `.default` is a `static var` currently that mimics the previous functionality).

Reply:
I definitely get the desire to make it clean at the call site, but I'm still wondering if an enum is the right choice since it sounds like a single `case` is enough to handle anything we throw at it. I believe we could get the same level of readability with a `custom(_:)` static-level function as well.
```swift
    struct URLRequestCreationStrategy {
        
        let creationBlock: (Request) -> URLRequest
        
        func urlRequest(using request: Request) -> URLRequest {
            return creationBlock(request)
        }
        
        public static func custom(_ creationBlock: @escaping (Request) -> URLRequest) -> URLRequestCreationStrategy {
            return URLRequestCreationStrategy(creationBlock: creationBlock)
        }
        
        public static var `default`: URLRequestCreationStrategy {
            return URLRequestCreationStrategy { request -> URLRequest in
                var urlRequest = URLRequest(url: request.url, cachePolicy: request.cachePolicy, timeoutInterval: request.timeout)
                urlRequest.httpMethod = request.method.rawValue
                urlRequest.httpBody = request.body?.data
                
                // Transform the headers from [HTTP.HeaderKey: HTTP.HeaderValue] to [String: String]
                let rawHeaders: [String: String] = Dictionary(uniqueKeysWithValues: (request.headers ?? [:]).map { ($0.rawValue, $1.rawValue) })
                urlRequest.allHTTPHeaderFields = rawHeaders
                
                return urlRequest
            }
        }
    }
```
I will admit that the enum approach is overall less code, thanks to the implicit storage of the associated value. For me, it's kind of a toss up between strictly adhering to the concept of an enum (2 or more) vs wanting to keep things as clean as possible. I'll leave the final decision up to you though.

Reply:
That's fair, when I originally set up I think I assumed I would need at least 2 cases to build off and then `static let/func` from there. The `struct` can be used with no changes to the call site, so I'm happy there.